src/main/app/controller/courts/AddressController.ts

```diff
@@ -0,0 +1,276 @@
+    }
+
+    return errors;
+  }
+
+  private validateAddressTypes(primary: DisplayAddress, secondary: DisplayAddress, writeToUsTypeId: number): string[] {
+    return writeToUsTypeId && !!primary.type_id && !!secondary.type_id && // validate only if types selected in both addresses
+    (this.addressFieldsNotEmpty(secondary) || secondary.postcode?.trim()) && // validate only if secondary has some fields entered
+    (primary.type_id !== writeToUsTypeId && secondary.type_id !== writeToUsTypeId) // at least 1 write address should exist
+      ? [this.multipleVisitAddressError]
+      : [];
+  }
+
+  private addressFieldsNotEmpty(courtAddress: DisplayAddress): boolean {
+    return !!courtAddress.address_lines?.trim() || !!courtAddress.address_lines_cy?.trim() ||
+      !!courtAddress.town?.trim() || !!courtAddress.town_cy?.trim();
+  }
+
+  private postcodeIsValidFormat(postcode: string): boolean {
+    const postcodeRegex = new RegExp('([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|' +
+      '(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))' +
+      '\\s?[0-9][A-Za-z]{2})', 'g');
+
+    const match = postcode.match(postcodeRegex);
+    return match?.length === 1 && match[0] === postcode;
+  }
+
+  private getAddressTypesForSelect(addressTypes: AddressType[], isPrimaryAddress: boolean): SelectItem[] {
+    const allAddressTypes = addressTypes.map((at: AddressType) => ({ value: at.id, text: at.name, selected: false }));
+
+    return isPrimaryAddress
+      ? allAddressTypes
+      : allAddressTypes.filter(at => at.text.toLowerCase() !== this.visitOrContactUsAddressType.toLowerCase());
+  }
+
+  private checkErrorResponseForPostcodeErrors(error: AxiosError, addresses: DisplayCourtAddresses):
+    { primaryInvalid: boolean; secondaryInvalid: boolean; errors: string[] } {
+
+    let primaryPostcodeInvalid = false;
+    let secondaryPostcodeInvalid = false;
+    const errors: string[] = [];
+
+    // We expect an array of invalid postcodes in the body of the response
+    if (Array.isArray(error.response.data)) {
+      const invalidPostcodes = error.response.data as string[];
+
+      invalidPostcodes.forEach(invalidPostcode => {
+        if (invalidPostcode === addresses.primary?.postcode) {
+          primaryPostcodeInvalid = true;
+          errors.push(this.primaryAddressPrefix + this.postcodeNotFoundError);
+        } else if (invalidPostcode === addresses.secondary?.postcode) {
+          secondaryPostcodeInvalid = true;
+          errors.push(this.secondaryAddressPrefix + this.postcodeNotFoundError);
+        }
+      });
+    }
+
+    return { primaryInvalid: primaryPostcodeInvalid, secondaryInvalid: secondaryPostcodeInvalid, errors: errors };
+  }
+
+  private convertToDisplayAddresses(addresses: CourtAddress[]): DisplayCourtAddresses {
+    const courtAddresses: DisplayCourtAddresses = { primary: null, secondary: null };
+    if (addresses.length > 0) {
```

Comment:
I assume from this that if the fact api returns zero addresses, that we will be returning a dictionary with two null values to the frontend? Which is fine

I will note that If we will always have one address, then I assume that we will not need the `if(address.length > 0)` condition however

Reply:
When testing I found that closed courts sometimes have zero addresses but still allow editing in the admin portal.